### ISSUE TITLE
[Feature Request] Better default font family that supports all unicode chars

### ISSUE THREAD

**Issue by tddschn** [matplotlib/matplotlib#20251]
### Problem

I'm working with a dataset that contains lots of east Asian characters. When trying to plot the data, I got a flood of warnings like this: `RuntimeWarning: Glyph <number> missing from current font.
  font.set_text(s, 0.0, flags=flags)`, and the generated plot had no readable texts.

Since I'm new to matplotlib and was calling `plot` from a pandas `DataFrame`, I couldn't find a trivial way to get the chars displayed correctly. Some suggest downloading a font file and configuring the font family with the path to that file, all seems quite complicated and I couldn't figure how to do that directly from `DataFrame.plot` arguments.

A few quick searches lead me to this issue from May 2019: https://github.com/matplotlib/matplotlib/issues/14269#issuecomment-493898037
> That is because the default font (DejaVu Sans) does not have CJKV

I don't think the current default font is a sensible choice since it only support latin characters.

### Proposed Solution

- Dynamically determine the default font family by choosing from a list of available fonts that supports unicode characters.
- Only fallback to font families with poor or no unicode support when no such fonts are available on users' system.
- Allow people to work with matplotlib with minimal friction when it comes to unicode text or data.

### Additional context and prior art

Warning screenshot: 
![image](https://user-images.githubusercontent.com/45612704/118611466-73a21180-b7ef-11eb-9895-e3bb7d66c3b4.png)

Plot generated:
![image](https://user-images.githubusercontent.com/45612704/118619214-f5e20400-b7f6-11eb-824e-c514e208bed6.png)

matplotlib version: 3.4.2 (from conda-forge)
OS: macOS 11.3.1

**Comment by jklymak** [comment]
Does https://matplotlib.org/stable/gallery/text_labels_and_annotations/font_family_rc_sgskip.html help?

This SO post is also quite good: https://stackoverflow.com/questions/45877746/changing-fonts-in-matplotlib

Google doesn't turn up more complete docs on our side, which seems a shame.  Are there more @aitikgupta ?

**Comment by tddschn** [comment]
@jklymak Thanks for the reply!
> Does matplotlib.org/stable/gallery/text_labels_and_annotations/font_family_rc_sgskip.html help?

I set `font.family` to `Hack Nerd Font`, and this was the output:
```python3
from matplotlib import rcParams
rcParams['font.family'] = 'Hack Nerd Font'
rcParams['font.sans-serif'] = ['Hack Nerd Font']
import matplotlib.pyplot as plt
fig, ax = plt.subplots()
ax.plot([1, 2, 3], label='test')

ax.legend()
ax.set_xlabel('测试')
plt.show()
```
![image](https://user-images.githubusercontent.com/45612704/118677699-044c1200-b82f-11eb-86d0-728a760a625c.png)

MPL didn't complain about missing glyphs, but the characters are not rendered correctly.  


> This SO post is also quite good: stackoverflow.com/questions/45877746/changing-fonts-in-matplotlib

I was able to get a list of fonts file (font_list) installed on my machine, but
The last line of the code 
```
font_names = [fm.FontProperties(fname=fname).get_name() for fname in font_list]

RuntimeError: In FT2Font: Could not set the fontsize (error code 0x17)
```
didn't work.

**Comment by aitikgupta** [comment]
> MPL didn't complain about missing glyphs, but the characters are not rendered correctly.

If there are missing glyphs (text not rendering correctly) - Matplotlib _is_ supposed to emit out the warnings, you might want to try it again?

Are you certain that the font you're using (Hack Nerd Font) contains these glyphs - `测试`? I tried reproducing with https://github.com/ryanoasis/nerd-fonts/tree/master/patched-fonts/Hack, it did warn me about the missing glyphs from this font too.

You can follow this [blog](https://scentellegher.github.io/visualization/2018/05/02/custom-fonts-matplotlib.html) to use custom fonts with Pandas, another [SO discussion](https://stackoverflow.com/questions/35668219/how-to-set-up-a-custom-font-with-custom-path-to-matplotlib-global-font) you might want to take a look at.

Basically there's 2 ways:
1. import matplotlib, set custom font as global font.
2. import matplotlib, define an `ax` with `font=/path/to/font.ttf`, and pass that to pandas.

**Comment by aitikgupta** [comment]
> Google doesn't turn up more complete docs on our side

Unfortunately, we don't have a solid entry point which talks in-depth about using custom fonts - we might be tracking this here: https://github.com/matplotlib/matplotlib/issues/5941

**Comment by jklymak** [comment]
Sure let's continue discussion at those issues. Thanks!

**Comment by tddschn** [comment]
@aitikgupta 
> Are you certain that the font you're using (Hack Nerd Font) contains these glyphs - 测试?

Yes, I believe that Hack Nerd Font has full unicode support :)
The output of these commands in the terminal:
```zsh
echo "测试"
neofetch term_font
```
<img src=https://user-images.githubusercontent.com/45612704/118742159-6d10aa00-b882-11eb-904a-7633ad6f0e18.png width=300px>

> I followed all the links but I don't think any of those methods would work, since the font I used is indeed globally installed and I use it in the Kitty terminal and VS Code
<img src=https://user-images.githubusercontent.com/45612704/118742750-95e56f00-b883-11eb-9d7f-770c3cc8e9ca.png width=500px>

**Comment by QuLogic** [comment]
Nerd Fonts simply adds a bunch of fancy symbols, so we can ignore that part of it. Going to the Hack fonts (https://sourcefoundry.org/hack/) details, they state:

> Over 1500 glyphs that include lovingly tuned extended Latin, modern Greek, and Cyrillic character sets.

That is nowhere near enough glyphs to cover CJK, and if you check it in e.g., fontforge, you will find no CJK glyphs. Your terminal supports glyph fallback and is picking a different font, as is discussed in the other issues.

**Comment by tddschn** [comment]
@QuLogic You're right. 

I found a working solution but it's ugly (PingFang is a system font that supports CJK chars shipped with macOS):
<img src=https://user-images.githubusercontent.com/45612704/118793234-fd72dd00-b8ca-11eb-92d2-b3e3532c5234.png width=550>

**Comment by tddschn** [comment]
@aitikgupta 
> Basically there's 2 ways:
import matplotlib, set custom font as global font.
import matplotlib, define an ax with custom font, and pass that to pandas.

Those are great ideas!
I tried to set the font family for the ax but with no success.  
Can you show me how to do both? Thanks!

**Comment by tacaswell** [comment]
Also see: https://matplotlib.org/stable/tutorials/text/text_props.html#text-with-non-latin-glyphs

**Comment by jklymak** [comment]
Oh good find. I knew we had something like that, but it's not easy to find.

**Comment by thedirk** [comment]
Any progress here? I'm having the same issue, my fonts display properly in my Pandas DataFrame, but if I set the font family, _only_ my non-standard characters display (and the Latin characters do not!).

**Comment by tacaswell** [comment]
@thedirk Yes, see https://matplotlib.org/stable/users/prev_whats_new/whats_new_3.6.0.html#font-fallback that lets you set up multiple fonts and we will fallback through them until we find the glyph requested.
